File: internal/server/jsonx.go
package server

import (
	"fmt"
	"strings"
	"unicode/utf16"
)

// JType — тип JSON-значения.
type JType byte

const (
	JObj   JType = 'o'
	JArr   JType = 'a'
	JStr   JType = 's'
	JNum   JType = 'n'
	JBool  JType = 'b'
	JNull  JType = 'z'
	JUndef JType = 'u'
)

// JValue — JSON-значение, сохраняющее порядок ключей (как JS-объект).
type JValue struct {
	Type JType
	// объект
	keys []string
	vals map[string]*JValue
	// массив
	items []*JValue
	// строка / число / bool (сырое слово)
	str string
	raw string
}

// Obj создаёт объект.
func Obj() *JValue { return &JValue{Type: JObj, vals: map[string]*JValue{}} }

// Arr создаёт массив.
func Arr() *JValue { return &JValue{Type: JArr} }

// Str создаёт строковое значение.
func Str(s string) *JValue { return &JValue{Type: JStr, str: s} }

// Num создаёт числовое значение (raw — исходное число).
func Num(raw string) *JValue { return &JValue{Type: JNum, raw: raw} }

// BoolVal создаёт булево значение.
func BoolVal(raw string) *JValue { return &JValue{Type: JBool, raw: raw} }

// Null создаёт null.
func Null() *JValue { return &JValue{Type: JNull} }

// Undef создаёт undefined (при сериализации в объекте пропускается).
func Undef() *JValue { return &JValue{Type: JUndef} }

// Set помещает ключ в объект.
func (v *JValue) Set(k string, val *JValue) {
	if v.keys == nil {
		v.keys = nil
	}
	v.keys = append(v.keys, k)
	v.vals[k] = val
}

// Get возвращает значение ключа объекта или nil.
func (v *JValue) Get(k string) *JValue {
	if v.Type != JObj {
		return nil
	}
	return v.vals[k]
}

// Keys возвращает ключи объекта в порядке вставки.
func (v *JValue) Keys() []string {
	if v.Type != JObj {
		return nil
	}
	return v.keys
}

// StrVal возвращает строковое значение (если это строка).
func (v *JValue) StrVal() (string, bool) {
	if v == nil || v.Type != JStr {
		return "", false
	}
	return v.str, true
}

// AsString возвращает строку или "" для не-строк.
func (v *JValue) AsString() string {
	if v == nil {
		return ""
	}
	if v.Type == JStr {
		return v.str
	}
	if v.Type == JNum || v.Type == JBool {
		return v.raw
	}
	return ""
}

// IsString сообщает, что значение — строка.
func (v *JValue) IsString() bool { return v != nil && v.Type == JStr }

// jParser — парсер JSON с сохранением порядка ключей.
type jParser struct {
	s   string
	pos int
}

func (p *jParser) skipWS() {
	for p.pos < len(p.s) {
		switch p.s[p.pos] {
		case ' ', '\t', '\n', '\r':
			p.pos++
		default:
			return
		}
	}
}

func (p *jParser) parseValue() (*JValue, error) {
	p.skipWS()
	if p.pos >= len(p.s) {
		return nil, fmt.Errorf("unexpected end")
	}
	switch p.s[p.pos] {
	case '{':
		return p.parseObject()
	case '[':
		return p.parseArray()
	case '"':
		sv, err := p.parseString()
		if err != nil {
			return nil, err
		}
		return Str(sv), nil
	case 't':
		if strings.HasPrefix(p.s[p.pos:], "true") {
			p.pos += 4
			return BoolVal("true"), nil
		}
	case 'f':
		if strings.HasPrefix(p.s[p.pos:], "false") {
			p.pos += 5
			return BoolVal("false"), nil
		}
	case 'n':
		if strings.HasPrefix(p.s[p.pos:], "null") {
			p.pos += 4
			return Null(), nil
		}
	default:
		// число
		start := p.pos
		for p.pos < len(p.s) {
			c := p.s[p.pos]
			if (c >= '0' && c <= '9') || c == '-' || c == '+' || c == '.' || c == 'e' || c == 'E' {
				p.pos++
			} else {
				break
			}
		}
		if p.pos > start {
			return Num(p.s[start:p.pos]), nil
		}
	}
	return nil, fmt.Errorf("unexpected char at %d", p.pos)
}

func (p *jParser) parseObject() (*JValue, error) {
	obj := Obj()
	p.pos++ // '{'
	for {
		p.skipWS()
		if p.pos >= len(p.s) {
			return nil, fmt.Errorf("unexpected end of object")
		}
		if p.s[p.pos] == '}' {
			p.pos++
			return obj, nil
		}
		k, err := p.parseString()
		if err != nil {
			return nil, err
		}
		p.skipWS()
		if p.pos >= len(p.s) || p.s[p.pos] != ':' {
			return nil, fmt.Errorf("expected ':'")
		}
		p.pos++
		v, err := p.parseValue()
		if err != nil {
			return nil, err
		}
		obj.Set(k, v)
		p.skipWS()
		if p.pos >= len(p.s) {
			return nil, fmt.Errorf("unexpected end of object")
		}
		if p.s[p.pos] == ',' {
			p.pos++
			continue
		}
		if p.s[p.pos] == '}' {
			p.pos++
			return obj, nil
		}
		return nil, fmt.Errorf("expected ',' or '}'")
	}
}

func (p *jParser) parseArray() (*JValue, error) {
	arr := Arr()
	p.pos++ // '['
	for {
		p.skipWS()
		if p.pos >= len(p.s) {
			return nil, fmt.Errorf("unexpected end of array")
		}
		if p.s[p.pos] == ']' {
			p.pos++
			return arr, nil
		}
		v, err := p.parseValue()
		if err != nil {
			return nil, err
		}
		arr.items = append(arr.items, v)
		p.skipWS()
		if p.pos >= len(p.s) {
			return nil, fmt.Errorf("unexpected end of array")
		}
		if p.s[p.pos] == ',' {
			p.pos++
			continue
		}
		if p.s[p.pos] == ']' {
			p.pos++
			return arr, nil
		}
		return nil, fmt.Errorf("expected ',' or ']'")
	}
}

func (p *jParser) parseString() (string, error) {
	if p.pos >= len(p.s) || p.s[p.pos] != '"' {
		return "", fmt.Errorf("expected string")
	}
	p.pos++
	var b strings.Builder
	for p.pos < len(p.s) {
		c := p.s[p.pos]
		if c == '"' {
			p.pos++
			return b.String(), nil
		}
		if c == '\\' {
			p.pos++
			if p.pos >= len(p.s) {
				return "", fmt.Errorf("bad escape")
			}
			e := p.s[p.pos]
			p.pos++
			switch e {
			case '"':
				b.WriteByte('"')
			case '\\':
				b.WriteByte('\\')
			case '/':
				b.WriteByte('/')
			case 'b':
				b.WriteByte('\b')
			case 'f':
				b.WriteByte('\f')
			case 'n':
				b.WriteByte('\n')
			case 'r':
				b.WriteByte('\r')
			case 't':
				b.WriteByte('\t')
			case 'u':
				if p.pos+4 > len(p.s) {
					return "", fmt.Errorf("bad unicode escape")
				}
				var r rune
				for _, hc := range p.s[p.pos : p.pos+4] {
					r *= 16
					switch {
					case hc >= '0' && hc <= '9':
						r += rune(hc - '0')
					case hc >= 'a' && hc <= 'f':
						r += rune(hc-'a') + 10
					case hc >= 'A' && hc <= 'F':
						r += rune(hc-'A') + 10
					default:
						return "", fmt.Errorf("bad unicode escape")
					}
				}
				p.pos += 4
				if utf16.IsSurrogate(r) && p.pos+6 <= len(p.s) && p.s[p.pos] == '\\' && p.s[p.pos+1] == 'u' {
					// surrogate pair
					var r2 rune
					for _, hc := range p.s[p.pos+2 : p.pos+6] {
						r2 *= 16
						switch {
						case hc >= '0' && hc <= '9':
							r2 += rune(hc - '0')
						case hc >= 'a' && hc <= 'f':
							r2 += rune(hc-'a') + 10
						case hc >= 'A' && hc <= 'F':
							r2 += rune(hc-'A') + 10
						}
					}
					if utf16.IsSurrogate(r2) {
						dec := utf16.DecodeRune(r, r2)
						b.WriteRune(dec)
						p.pos += 6
						continue
					}
				}
				b.WriteRune(r)
			default:
				return "", fmt.Errorf("bad escape %q", e)
			}
			continue
		}
		b.WriteByte(c)
		p.pos++
	}
	return "", fmt.Errorf("unterminated string")
}

// ParseJSON разбирает JSON, сохраняя порядок ключей.
func ParseJSON(s string) (*JValue, error) {
	p := &jParser{s: s}
	v, err := p.parseValue()
	if err != nil {
		return nil, err
	}
	return v, nil
}

// jsEscapeStr экранирует строку как JSON.stringify.
func jsEscapeStr(s string, b *strings.Builder) {
	for _, r := range s {
		switch r {
		case '"':
			b.WriteString(`\"`)
		case '\\':
			b.WriteString(`\\`)
		case '\b':
			b.WriteString(`\b`)
		case '\f':
			b.WriteString(`\f`)
		case '\n':
			b.WriteString(`\n`)
		case '\r':
			b.WriteString(`\r`)
		case '\t':
			b.WriteString(`\t`)
		default:
			if r < 0x20 {
				b.WriteString(fmt.Sprintf(`\u%04x`, r))
			} else {
				b.WriteRune(r)
			}
		}
	}
}

// JStringify сериализует значение как JSON.stringify (порядок ключей сохраняется).
func JStringify(v *JValue) string {
	var b strings.Builder
	jWrite(&b, v)
	return b.String()
}

func jWrite(b *strings.Builder, v *JValue) {
	if v == nil || v.Type == JNull {
		b.WriteString("null")
		return
	}
	if v.Type == JUndef {
		b.WriteString("null")
		return
	}
	switch v.Type {
	case JStr:
		b.WriteByte('"')
		jsEscapeStr(v.str, b)
		b.WriteByte('"')
	case JNum, JBool:
		b.WriteString(v.raw)
	case JArr:
		b.WriteByte('[')
		for i, it := range v.items {
			if i > 0 {
				b.WriteByte(',')
			}
			jWrite(b, it)
		}
		b.WriteByte(']')
	case JObj:
		b.WriteByte('{')
		first := true
		for _, k := range v.keys {
			val := v.vals[k]
			if val != nil && val.Type == JUndef {
				continue
			}
			if !first {
				b.WriteByte(',')
			}
			first = false
			b.WriteByte('"')
			jsEscapeStr(k, b)
			b.WriteString(`":`)
			jWrite(b, val)
		}
		b.WriteByte('}')
	}
}

File: fioincline/jsonx.py
"""JSON-значение с сохранением порядка ключей и JS-экранированием (порт jsonx.go)."""

import json as _json

# Типы: obj, arr, str, num, bool, null, undef
J_OBJ = "o"
J_ARR = "a"
J_STR = "s"
J_NUM = "n"
J_BOOL = "b"
J_NULL = "z"
J_UNDEF = "u"


class JValue:
    __slots__ = ("type", "keys", "vals", "str", "raw", "items")

    def __init__(self, type_=J_UNDEF):
        self.type = type_
        self.keys = []      # порядок ключей объекта
        self.vals = {}      # dict ключ->JValue
        self.str = ""       # для строк
        self.raw = ""       # для num/bool
        self.items = []     # для массивов

    def set(self, k, val):
        self.keys.append(k)
        self.vals[k] = val

    def get(self, k):
        if self.type != J_OBJ:
            return None
        return self.vals.get(k)

    def is_string(self):
        return self.type == J_STR

    def str_val(self):
        if self.type == J_STR:
            return self.str, True
        return "", False


def obj():
    return JValue(J_OBJ)


def arr():
    return JValue(J_ARR)


def Str(s):
    v = JValue(J_STR)
    v.str = s
    return v


def Num(raw):
    v = JValue(J_NUM)
    v.raw = raw
    return v


def BoolVal(raw):
    v = JValue(J_BOOL)
    v.raw = raw
    return v


def Null():
    return JValue(J_NULL)


def Undef():
    return JValue(J_UNDEF)


def _escape(b, s):
    for r in s:
        if r == '"':
            b.append('\\"')
        elif r == "\\":
            b.append("\\\\")
        elif r == "\b":
            b.append("\\b")
        elif r == "\f":
            b.append("\\f")
        elif r == "\n":
            b.append("\\n")
        elif r == "\r":
            b.append("\\r")
        elif r == "\t":
            b.append("\\t")
        elif ord(r) < 0x20:
            b.append("\\u%04x" % ord(r))
        else:
            b.append(r)


def _write(b, v):
    if v is None or v.type in (J_NULL, J_UNDEF):
        b.append("null")
        return
    t = v.type
    if t == J_STR:
        b.append('"')
        _escape(b, v.str)
        b.append('"')
    elif t in (J_NUM, J_BOOL):
        b.append(v.raw)
    elif t == J_ARR:
        b.append("[")
        for i, it in enumerate(v.items):
            if i > 0:
                b.append(",")
            _write(b, it)
        b.append("]")
    elif t == J_OBJ:
        b.append("{")
        first = True
        for k in v.keys:
            val = v.vals[k]
            if val is not None and val.type == J_UNDEF:
                continue
            if not first:
                b.append(",")
            first = False
            b.append('"')
            _escape(b, k)
            b.append('":')
            _write(b, val)
        b.append("}")


def j_stringify(v):
    b = []
    _write(b, v)
    return "".join(b)


class _Parser:
    def __init__(self, s):
        self.s = s
        self.pos = 0

    def skip_ws(self):
        while self.pos < len(self.s) and self.s[self.pos] in " \t\n\r":
            self.pos += 1

    def parse_value(self):
        self.skip_ws()
        if self.pos >= len(self.s):
            raise ValueError("unexpected end")
        c = self.s[self.pos]
        if c == "{":
            return self.parse_object()
        if c == "[":
            return self.parse_array()
        if c == '"':
            return Str(self.parse_string())
        if self.s.startswith("true", self.pos):
            self.pos += 4
            return BoolVal("true")
        if self.s.startswith("false", self.pos):
            self.pos += 5
            return BoolVal("false")
        if self.s.startswith("null", self.pos):
            self.pos += 4
            return Null()
        # число
        start = self.pos
        while self.pos < len(self.s) and self.s[self.pos] in "0123456789-+.eE":
            self.pos += 1
        if self.pos > start:
            return Num(self.s[start:self.pos])
        raise ValueError(f"unexpected char at {self.pos}")

    def parse_object(self):
        v = obj()
        self.pos += 1  # '{'
        while True:
            self.skip_ws()
            if self.pos >= len(self.s):
                raise ValueError("unexpected end of object")
            if self.s[self.pos] == "}":
                self.pos += 1
                return v
            if self.s[self.pos] != '"':
                raise ValueError("expected string key")
            k = self.parse_string()
            self.skip_ws()
            if self.pos >= len(self.s) or self.s[self.pos] != ":":
                raise ValueError("expected ':'")
            self.pos += 1
            val = self.parse_value()
            v.set(k, val)
            self.skip_ws()
            if self.pos >= len(self.s):
                raise ValueError("unexpected end of object")
            if self.s[self.pos] == ",":
                self.pos += 1
                continue
            if self.s[self.pos] == "}":
                self.pos += 1
                return v
            raise ValueError("expected ',' or '}'")

    def parse_array(self):
        v = arr()
        self.pos += 1  # '['
        while True:
            self.skip_ws()
            if self.pos >= len(self.s):
                raise ValueError("unexpected end of array")
            if self.s[self.pos] == "]":
                self.pos += 1
                return v
            item = self.parse_value()
            v.items.append(item)
            self.skip_ws()
            if self.pos >= len(self.s):
                raise ValueError("unexpected end of array")
            if self.s[self.pos] == ",":
                self.pos += 1
                continue
            if self.s[self.pos] == "]":
                self.pos += 1
                return v
            raise ValueError("expected ',' or ']'")

    def parse_string(self):
        if self.pos >= len(self.s) or self.s[self.pos] != '"':
            raise ValueError("expected string")
        self.pos += 1
        out = []
        while self.pos < len(self.s):
            c = self.s[self.pos]
            if c == '"':
                self.pos += 1
                return "".join(out)
            if c == "\\":
                self.pos += 1
                if self.pos >= len(self.s):
                    raise ValueError("bad escape")
                e = self.s[self.pos]
                self.pos += 1
                if e == '"':
                    out.append('"')
                elif e == "\\":
                    out.append("\\")
                elif e == "/":
                    out.append("/")
                elif e == "b":
                    out.append("\b")
                elif e == "f":
                    out.append("\f")
                elif e == "n":
                    out.append("\n")
                elif e == "r":
                    out.append("\r")
                elif e == "t":
                    out.append("\t")
                elif e == "u":
                    hexstr = self.s[self.pos:self.pos + 4]
                    if len(hexstr) != 4:
                        raise ValueError("bad unicode escape")
                    try:
                        out.append(chr(int(hexstr, 16)))
                    except ValueError:
                        raise ValueError("bad unicode escape")
                    self.pos += 4
                else:
                    raise ValueError(f"bad escape {e}")
                continue
            out.append(c)
            self.pos += 1
        raise ValueError("unterminated string")


def parse_json(s):
    return _Parser(s).parse_value()


# Парсинг с сохранением порядка через стандартный json -> JValue (обратная совместимость).
def from_py(data):
    """Переводит Python-объект (dict/list/...) в JValue с сохранением порядка."""
    if data is None:
        return Null()
    if isinstance(data, dict):
        v = obj()
        for k, val in data.items():
            v.set(str(k), from_py(val))
        return v
    if isinstance(data, list):
        v = arr()
        v.items = [from_py(x) for x in data]
        return v
    if isinstance(data, bool):
        return BoolVal("true" if data else "false")
    if isinstance(data, (int, float)):
        return Num(_json.dumps(data))
    return Str(str(data))
